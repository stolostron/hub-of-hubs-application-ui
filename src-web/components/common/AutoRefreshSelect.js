// Copyright Contributors to the Open Cluster Management project
'use strict'

import _ from 'lodash'

import React, { Component, Fragment } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { withRouter } from 'react-router-dom'
import {
  AcmAutoRefreshSelect,
  AcmRefreshTime
} from '@stolostron/ui-components'

import { getSelectedId } from './QuerySwitcher'
import {
  INITIAL_REFRESH_TIME,
  RESOURCE_TYPES,
  REFRESH_TIMES,
  LOCAL_HUB_NAME
} from '../../../lib/shared/constants'
import { fetchResources } from '../../actions/common'
import { combineFilters } from '../../actions/filters'
import { fetchTopology } from '../../actions/topology'

const getResourceTypeForLocation = location => {
  const defaultOption = location.pathname.includes(
    'multicloud/applications/advanced'
  )
    ? 'subscriptions'
    : 'applications'

  const options = [
    { id: 'applications', resourceType: RESOURCE_TYPES.QUERY_APPLICATIONS },
    {
      id: 'subscriptions',
      resourceType: RESOURCE_TYPES.QUERY_SUBSCRIPTIONS
    },
    {
      id: 'placementrules',
      resourceType: RESOURCE_TYPES.QUERY_PLACEMENTRULES
    },
    { id: 'channels', resourceType: RESOURCE_TYPES.QUERY_CHANNELS }
  ]

  const selectedId = getSelectedId({ location, options, defaultOption })
  const resourceType = options.find(o => o.id === selectedId).resourceType
  const routePaths = location ? _.get(location, 'pathname', '').split('/') : []

  if (defaultOption === 'applications' && routePaths.length === 5) {
    //this is the single app page
    return RESOURCE_TYPES.HCM_APPLICATIONS
  } else {
    //this is one of the resource tables
    return resourceType
  }
}

class AutoRefreshSelect extends Component {
  render() {
    const {
      fetchTableResources,
      fetchAppTopology,
      status,
      isEditTab
    } = this.props

    const refetch = () => {
      // skip refresh, if the page is still refreshing from the previous call, if this is the initial load page
      // or if it's less then 5s from the previous fetch
      const deltaSinceLastUpdate = this.props.resourceRefreshTime
        ? Date.now() - this.props.resourceRefreshTime
        : 6000
      if (
        !_.includes(['IN_PROGRESS', 'INCEPTION'], this.props.status) &&
        deltaSinceLastUpdate > 5000
      ) {
        if (this.props.resourceType === RESOURCE_TYPES.HCM_APPLICATIONS) {
          const search = window.location.search
          const searchItems = search ? new URLSearchParams(search) : undefined
          let cluster = LOCAL_HUB_NAME
          let apiVersion = 'app.k8s.io/v1beta1'
          if (searchItems && searchItems.get('apiVersion')) {
            apiVersion = searchItems.get('apiVersion')
          }
          if (searchItems && searchItems.get('cluster')) {
            cluster = searchItems.get('cluster')
          }
          const name = _.get(this.props.fetchFilters, 'application.name', '')
          const namespace = _.get(
            this.props.fetchFilters,
            'application.namespace',
            ''
          )
          const channel = _.get(
            this.props.fetchFilters,
            'application.channel',
            ''
          )
          const filterWithLocation = {
            application: {
              name,
              namespace,
              channel,
              apiVersion,
              cluster
            }
          }
          fetchAppTopology(filterWithLocation, true) //fetch single app
        } else {
          fetchTableResources(this.props.resourceType, []) //fetch table resources
        }
      }
    }

    return (
      <Fragment>
        {!isEditTab && (
          <div className="auto-refresh">
            <AcmAutoRefreshSelect
              refetch={refetch}
              refreshIntervals={REFRESH_TIMES}
              initPollInterval={INITIAL_REFRESH_TIME}
            />
            <AcmRefreshTime
              timestamp={this.props.resourceRefreshTime || Date.now()}
              reloading={status !== 'DONE'}
            />
          </div>
        )}
      </Fragment>
    )
  }
}

const mapStateToProps = (state, ownProps) => {
  const location = ownProps.route
  const resourceType = getResourceTypeForLocation(location)
  const routePaths = location ? _.get(location, 'pathname', '').split('/') : []
  const isEditTab =
    routePaths.length === 6 || //edit app
    (routePaths.length === 4 &&
      ['create' || 'argoappset'].indexOf(routePaths[3]) !== -1) //new app or argoappset
  if (isEditTab) {
    //this is the editor tab, the refresh action is not showing here
    return { isEditTab: true }
  }

  let fetchFilters = _.get(state, 'topology.fetchFilters', {})
  if (resourceType === RESOURCE_TYPES.HCM_APPLICATIONS) {
    //make sure the topology stored by state matches the application route
    //if going to the editor from the app table then switch to the Topology tab,
    //the status returns the topology displayed previously by the topology tab
    const name = routePaths[4]
    const namespace = routePaths[3]
    if (
      _.get(state, 'topology.activeFilters.application.name', '') !== name ||
      _.get(state, 'topology.activeFilters.application.namespace', '') !==
        namespace
    ) {
      //need to rebase the state topology to the router app
      fetchFilters = { application: { name, namespace, channel: '' } }
    }
  }
  const typeListName = resourceType.list
  const status = state[typeListName].status
  return {
    resourceRefreshTime: _.get(state[typeListName], 'responseTime'),
    resourceType,
    status,
    fetchFilters,
    forceReload: state[typeListName].forceReload,
    isEditTab: false
  }
}

const mapDispatchToProps = dispatch => {
  const search = window.location.search
  const searchItems = search ? new URLSearchParams(search) : undefined
  let cluster
  let apiVersion = 'app.k8s.io/v1beta1'
  if (searchItems && searchItems.get('apiVersion')) {
    apiVersion = searchItems.get('apiVersion')
  }
  if (searchItems && searchItems.get('cluster')) {
    cluster = searchItems.get('cluster')
  }
  return {
    fetchTableResources: (resourceType, selectedFilters) => {
      dispatch(fetchResources(resourceType, combineFilters(selectedFilters)))
    },
    fetchAppTopology: (fetchFilters, reloading) => {
      const name = _.get(fetchFilters, 'application.name', '')
      const namespace = _.get(fetchFilters, 'application.namespace', '')
      const channel = _.get(fetchFilters, 'application.channel', '')
      const filterApiVersion =
        _.get(fetchFilters, 'application.apiVersion') || apiVersion
      const filterCluster =
        _.get(fetchFilters, 'application.cluster') || cluster
      const filterWithLocation = {
        application: {
          name,
          namespace,
          channel,
          apiVersion: filterApiVersion,
          cluster: filterCluster
        }
      }
      dispatch(
        fetchTopology(
          { filter: { ...filterWithLocation } },
          filterWithLocation,
          reloading
        )
      )
    }
  }
}

AutoRefreshSelect.propTypes = {
  fetchAppTopology: PropTypes.func,
  fetchFilters: PropTypes.object,
  fetchTableResources: PropTypes.func,
  isEditTab: PropTypes.bool,
  resourceRefreshTime: PropTypes.number,
  resourceType: PropTypes.object,
  status: PropTypes.string
}

export default withRouter(
  connect(mapStateToProps, mapDispatchToProps)(AutoRefreshSelect)
)
