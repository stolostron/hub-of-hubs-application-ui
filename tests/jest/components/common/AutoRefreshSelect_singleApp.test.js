// Copyright (c) 2021 Red Hat, Inc.
jest.mock("../../../../lib/client/apollo-client", () => ({
  get: jest.fn(resourceType => {
    //resourceType.list is always ApplicationsList
    return Promise.resolve({
      data: {
        applications: {
          "mortgage-app-default": {
            _uid: "local-cluster/5cd1d4c7-52aa-11ea-bf05-00000a102d26",
            name: "mortgage-app",
            namespace: "default",
            dashboard:
              "https://localhost:443/grafana/dashboard/db/mortgage-app-dashboard-via-federated-prometheus?namespace=default",
            clusterCount: 1,
            hubSubscriptions: [
              {
                _uid: "local-cluster/5cdc0d8d-52aa-11ea-bf05-00000a102d26",
                status: "Propagated",
                channel: "default/mortgage-channel",
                __typename: "Subscription"
              }
            ],
            created: "2018-02-18T23:57:04Z",
            __typename: "Application"
          }
        }
      }
    });
  }),
  search: jest.fn((searchQuery, searchInput) => {
    return Promise.resolve({
      data: {
        applications: {
          "mortgage-app-default": {
            _uid: "local-cluster/5cd1d4c7-52aa-11ea-bf05-00000a102d26",
            name: "mortgage-app",
            namespace: "default",
            dashboard:
              "https://localhost:443/grafana/dashboard/db/mortgage-app-dashboard-via-federated-prometheus?namespace=default",
            clusterCount: 1,
            hubSubscriptions: [
              {
                _uid: "local-cluster/5cdc0d8d-52aa-11ea-bf05-00000a102d26",
                status: "Propagated",
                channel: "default/mortgage-channel",
                __typename: "Subscription"
              }
            ],
            created: "2018-02-18T23:57:04Z",
            __typename: "Application"
          }
        }
      }
    });
  }),
  getResource: jest.fn((resourceType, variables) => {
    return Promise.resolve({
      data: {
        applications: {
          "mortgage-app-default": {
            _uid: "local-cluster/5cd1d4c7-52aa-11ea-bf05-00000a102d26",
            name: "mortgage-app",
            namespace: "default",
            dashboard:
              "https://localhost:443/grafana/dashboard/db/mortgage-app-dashboard-via-federated-prometheus?namespace=default",
            clusterCount: 1,
            hubSubscriptions: [
              {
                _uid: "local-cluster/5cdc0d8d-52aa-11ea-bf05-00000a102d26",
                status: "Propagated",
                channel: "default/mortgage-channel",
                __typename: "Subscription"
              }
            ],
            created: "2018-02-18T23:57:04Z",
            __typename: "Application"
          }
        }
      }
    });
  })
}));

("use strict");
import React from "react";
import AutoRefreshSelect from "../../../../src-web/components/common/AutoRefreshSelect";
import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";

import thunkMiddleware from "redux-thunk";
import configureMockStore from "redux-mock-store";

import { reduxStoreAppPipelineWithCEM } from "../TestingData";

const middleware = [thunkMiddleware];
const mockStore = configureMockStore(middleware);
const storeSingleApp = mockStore(reduxStoreAppPipelineWithCEM);

const locationApp = {
  pathname: "/multicloud/applications/helloworld-ns/helloworld"
};

const dateNowStub = jest.fn(() => 1530518207007);
global.Date.now = dateNowStub;

describe("AutoRefreshSelect on single app", () => {
  it("renders correctly when visible on the single app page", () => {
    const wrapper = mount(
      <BrowserRouter>
        <Provider store={storeSingleApp}>
          <AutoRefreshSelect route={locationApp} />
        </Provider>
      </BrowserRouter>
    );

    wrapper
      .find("#refresh-toggle")
      .at(0)
      .simulate("click");

    wrapper.find(".pf-c-dropdown__toggle-text").simulate("click");

    wrapper
      .find("#refresh-toggle")
      .at(0)
      .simulate("click");

    wrapper
      .find("#refresh-disable")
      .at(0)
      .simulate("click");

    expect(toJson(wrapper.render())).toMatchSnapshot();
  });
});
