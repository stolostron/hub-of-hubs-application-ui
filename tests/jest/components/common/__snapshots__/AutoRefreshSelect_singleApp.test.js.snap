// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutoRefreshSelect on single app renders correctly when visible on the single app page 1`] = `
<div
  class="auto-refresh"
>
  <div
    class="makeStyles-container-1"
  >
    <div
      aria-label="refresh-icon"
      class="makeStyles-reloadButton-2"
      id="refresh-icon"
      role="button"
      tabindex="0"
    >
      <svg
        aria-hidden="true"
        class="makeStyles-icon-4"
        fill="currentColor"
        height="1em"
        role="img"
        style="vertical-align: -0.125em;"
        viewBox="0 0 512 512"
        width="1em"
      >
        <path
          d="M370.72 133.28C339.458 104.008 298.888 87.962 255.848 88c-77.458.068-144.328 53.178-162.791 126.85-1.344 5.363-6.122 9.15-11.651 9.15H24.103c-7.498 0-13.194-6.807-11.807-14.176C33.933 94.924 134.813 8 256 8c66.448 0 126.791 26.136 171.315 68.685L463.03 40.97C478.149 25.851 504 36.559 504 57.941V192c0 13.255-10.745 24-24 24H345.941c-21.382 0-32.09-25.851-16.971-40.971l41.75-41.749zM32 296h134.059c21.382 0 32.09 25.851 16.971 40.971l-41.75 41.75c31.262 29.273 71.835 45.319 114.876 45.28 77.418-.07 144.315-53.144 162.787-126.849 1.344-5.363 6.122-9.15 11.651-9.15h57.304c7.498 0 13.194 6.807 11.807 14.176C478.067 417.076 377.187 504 256 504c-66.448 0-126.791-26.136-171.315-68.685L48.97 471.03C33.851 486.149 8 475.441 8 454.059V320c0-13.255 10.745-24 24-24z"
        />
      </svg>
    </div>
    <div
      aria-label="refetch-intervals"
      class="pf-c-dropdown makeStyles-buttonTitle-3"
      data-ouia-component-id="OUIA-Generated-Dropdown-1"
      data-ouia-component-type="PF4/Dropdown"
      data-ouia-safe="true"
      id="refresh-dropdown"
    >
      <button
        aria-expanded="false"
        aria-haspopup="true"
        aria-label="refresh-label"
        class="pf-c-dropdown__toggle pf-m-plain"
        data-ouia-component-id="OUIA-Generated-DropdownToggle-1"
        data-ouia-component-type="PF4/DropdownToggle"
        data-ouia-safe="true"
        id="refresh-toggle"
        type="button"
      >
        <span
          class="pf-c-dropdown__toggle-text"
        >
          Disable refresh
        </span>
        <span
          class="pf-c-dropdown__toggle-icon"
        >
          <svg
            aria-hidden="true"
            fill="currentColor"
            height="1em"
            role="img"
            style="vertical-align: -0.125em;"
            viewBox="0 0 320 512"
            width="1em"
          >
            <path
              d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
  <div
    class="makeStyles-timestamp-5"
  >
    <p>
      Last update: 3:56:32 AM
    </p>
  </div>
</div>
`;
